### PR TITLE
Hide underlying array in TArray

### PR DESCRIPTION
**Pull request by mijicd**
The array was public val, therefore accessible from outside.

### CODE REVIEW

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 1]
```diff
@@ -17,138 +17,193 @@
 package zio.stm
 
 /** Wraps array of [[TRef]] and adds methods for convenience.
- * Caution: most of methods are not stack-safe.
- * */
-class TArray[A] private (val array: Array[TRef[A]]) extends AnyVal {
+ *
+ */
+class TArray[A] private (private val array: Array[TRef[A]]) extends AnyVal {
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: Should we make this and the other `STM` types `final`? Technically classes extending `AnyVal` are already final but could be good to clarify intent. Looks like we do this elsewhere in the code base.

Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Absolutely agreed, done. Regarding the other "leakings", I'll tackle them in a separate PR(s), just to preserve the scope of this one.

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 2]
```diff
@@ -16,139 +16,193 @@
...
       }
 
-  /** Find the last element in the array matching a predicate. */
-  final def findLast(predicate: A => Boolean): STM[Nothing, Option[A]] =
-    new TArray(array.reverse).find(predicate)
+  /**
+   * Find the last element in the array matching a predicate.
+   */
+  final def findLast(p: A => Boolean): STM[Nothing, Option[A]] =
+    new TArray(array.reverse).find(p)
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: This is counter-intuitive, you would expect subsequent operations to be done on the same array, but here a new one is created, which could yield different results on those operations.

Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: This one's actually the old one (diff due to renaming). Technically, it is on the same one (TArray being value class), however, I have similar concerns on a couple more methods (e.g. `existsM`).

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 3]
```diff
@@ -16,139 +16,193 @@
...
+          else new TArray(array.tail).findM(p)
         }
       }
 
-  /** The first entry of the array, if it exists. */
-  final def firstOption: STM[Nothing, Option[A]] = if (array.isEmpty) STM.succeed(None) else array.head.get.map(Some(_))
+  /**
+   * The first entry of the array, if it exists.
+   */
+  final def firstOption: STM[Nothing, Option[A]] =
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: Note that we could add a `first` that just retries if the array is empty 🤔

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 2 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Edit: I've re-read what you wrote, and yes, this is even more extreme actually.

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 3 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Oh, I believe that was intentionally removed as we didn't want to expose "unsafe" methods (see #2446).

Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Yes, my point is thanks to the semantics of STM, you can make a safe version of `first` that just parks the computation if the array is empty 😄 WDYT?

Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: True that. But the question is should we reflect that behavior to all accesses, e.g. non-existing index?

Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: hmm, good question. I wouldn't, e.g. `array(-1)` would never make sense.

Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Ah, the thing is, the current API doesn't really allow elements to be added, so retrying won't make much sense.

Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Ah right 👍

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 4]
```diff
@@ -16,139 +16,193 @@
...
+    if (array.isEmpty) STM.succeed(None) else array.head.get.map(Some(_))
 
-  /** Atomically folds [[TArray]] with pure function. */
+  /**
+   * Atomically folds [[TArray]] with pure function.
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: typo: a pure function

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 5]
```diff
@@ -16,139 +16,193 @@
...
-    countM(predicate).map(_ == array.length) //Universal quantification should not be order dependent so the entire array needs to be observed.
+  /**
+   * Atomically evaluate the conjunction of an STM predicate across the members of the array.
+   */
+  final def forallM[E](p: A => STM[E, Boolean]): STM[E, Boolean] =
+    countM(p).map(_ == array.length)
 
-  /** Atomically performs side-effect for each item in array */
+  /**
+   * Atomically performs side-effect for each item in array.
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: perform a transaction for each item in the array?

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 6]
```diff
@@ -16,139 +16,193 @@
...
+   */
   final def foreach[E](f: A => STM[E, Unit]): STM[E, Unit] =
-    this.foldM(())((_, a) => f(a))
+    foldM(())((_, a) => f(a))
 
-  /** Get the first index of a specific value in the array or -1 if it does not occur. */
+  /**
+   * Get the first index of a specific value in the array or -1 if it does not occur.
+   */
   final def indexOf(a: A): STM[Nothing, Int] = indexOf(a, 0)
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: Should we push this failure mode into the error channel?

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 7]
```diff
@@ -157,55 +211,72 @@ class TArray[A] private (val array: Array[TRef[A]]) extends AnyVal {
...
+    foldM[E, Option[A]](None) { (optAcc, a) =>
+      optAcc match {
+        case Some(acc) => op(acc, a).map(Some(_))
+        case _         => STM.succeed(Some(a))
+      }
+    }
 
-  /** Atomically updates all [[TRef]]s inside this array using pure function. */
+  /**
+   * Atomically updates all elements using pure function.
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: typo: a pure function

File: core/shared/src/main/scala/zio/stm/TArray.scala [hunk 8]
```diff
@@ -157,55 +211,72 @@ class TArray[A] private (val array: Array[TRef[A]]) extends AnyVal {
...
+   */
   final def update(index: Int, fn: A => A): STM[Nothing, A] =
     if (0 <= index && index < array.length) array(index).update(fn)
     else STM.die(new ArrayIndexOutOfBoundsException(index))
 
-  /** Atomically updates element in the array with given transactional effect. */
+  /**
+   * Atomically updates element in the array with given transactional effect.
+   */
   final def updateM[E](index: Int, fn: A => STM[E, A]): STM[E, A] =
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/stm/TArray.scala]: for other parts of zio we have decided to use M suffix only for function involving the zio datatype.
How about we rename these to updateTransactional updateSTM for consistency?

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 1 above)
Comment (saraiva132) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: @mijicd This could also potentially be good first PRs. :D

Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: @saraiva132 I'll double-check if it makes sense to do such change, and if so, I'll make some issues!

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 4 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Done.

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 5 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Slightly different, but done.

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 7 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: Done.

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 8 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: I'd leave it for now. We can deprecate this in the future, assuming there's an agreement about that, although, I find the current API to be more convenient.

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 6 above)
Comment (mijicd) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: I don't think we should as the goal was for this method to be symmetric with the one in `Array`.

Comment (mschuwalow) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: ok, makes sense

File: core/shared/src/main/scala/zio/stm/TArray.scala (hunk 8 above)
Comment (mschuwalow) [reply core/shared/src/main/scala/zio/stm/TArray.scala]: 👍

**Pull request merged**